README.md
# Chef Handler SNS
[![Gem Version](http://img.shields.io/gem/v/chef-handler-sns.svg?style=flat)](http://badge.fury.io/rb/chef-handler-sns)
[![Documentation](http://img.shields.io/badge/docs-rdoc.info-blue.svg?style=flat)](http://www.rubydoc.info/gems/chef-handler-sns)
[![GitHub](http://img.shields.io/badge/github-zuazo/chef--handler--sns-blue.svg?style=flat)](https://github.com/zuazo/chef-handler-sns)
[![License](https://img.shields.io/github/license/zuazo/chef-handler-sns.svg?style=flat)](#license-and-author)

[![Dependency Status](http://img.shields.io/gemnasium/zuazo/chef-handler-sns.svg?style=flat)](https://gemnasium.com/zuazo/chef-handler-sns)
[![Code Climate](http://img.shields.io/codeclimate/github/zuazo/chef-handler-sns.svg?style=flat)](https://codeclimate.com/github/zuazo/chef-handler-sns)
[![Build Status](http://img.shields.io/travis/zuazo/chef-handler-sns.svg?style=flat)](https://travis-ci.org/zuazo/chef-handler-sns)
[![Coverage Status](http://img.shields.io/coveralls/zuazo/chef-handler-sns.svg?style=flat)](https://coveralls.io/r/zuazo/chef-handler-sns?branch=master)
[![Inline docs](http://inch-ci.org/github/zuazo/chef-handler-sns.svg?branch=master&style=flat)](http://inch-ci.org/github/zuazo/chef-handler-sns)

A simple Chef report handler that reports status of a Chef run through [Amazon SNS](http://aws.amazon.com/sns/), [including IAM roles support](#usage-with-amazon-iam-roles).

[Amazon SNS](http://aws.amazon.com/sns/) can send notifications by SMS, email, [Amazon SQS](http://aws.amazon.com/sqs/) queues or to any HTTP endpoint.

We recommend using the [`chef_handler_sns` cookbook](https://supermarket.chef.io/cookbooks/chef_handler_sns) for easy installation.

This Chef Handler is heavily based on [Joshua Timberman](https://github.com/jtimberman) examples.

* https://docs.chef.io/handlers.html#exception-report-handlers

## Requirements

* Amazon AWS: uses Amazon SNS service.
* Ruby `2` or higher (recommended `2.1` or higher).

## Usage

You can install this handler in two ways:

### Method 1: in the Chef Config File

You can install the RubyGem and configure Chef to use it:

    $ gem install chef-handler-sns

Then add to the configuration (`/etc/chef/solo.rb` for chef-solo or `/etc/chef/client.rb` for chef-client):

```ruby
require 'chef/handler/sns'

# Create the handler
sns_handler = Chef::Handler::Sns.new

# Your Amazon AWS credentials
sns_handler.access_key '***AMAZON-KEY***'
sns_handler.secret_key '***AMAZON-SECRET***'

# Some Amazon SNS configurations
sns_handler.topic_arn 'arn:aws:sns:***'
sns_handler.region 'us-east-1' # optional

# Add your handler
exception_handlers << sns_handler
```

### Method 2: in a Recipe with the `chef_handler` LWRP

**Note:** This method will not catch errors before the convergence phase. Use the previous method if you want to be able to report such errors.

Use the [`chef_handler` LWRP](https://supermarket.chef.io/cookbooks/chef_handler), creating a recipe with the following:

```ruby
# Handler configuration options
argument_array = [
  access_key: '***AMAZON-KEY***',
  secret_key: '***AMAZON-SECRET***',
  topic_arn: 'arn:aws:sns:***'
]

# Install the `chef-handler-sns` RubyGem during the compile phase
chef_gem 'chef-handler-sns' do
  compile_time true # Only for Chef 12
end

# Then activate the handler with the `chef_handler` LWRP
chef_handler 'Chef::Handler::Sns' do
  source 'chef/handler/sns'
  arguments argument_array
  supports exception: true
  action :enable
end
```

See the [`chef_handler_sns` cookbook provider code](https://github.com/zuazo/chef_handler_sns-cookbook/blob/master/providers/default.rb) for a more complete working example.

### Method 3: Using the `chef-client` Cookbook

You can also use the `node['chef_client']['config']` attribute of the [`chef-client`](https://github.com/chef-cookbooks/chef-client/tree/v4.3.2#start-report-exception-handlers) cookbook:

```ruby
node.default['chef_client']['config']['exception_handlers'] = [{
  'class' => 'Chef::Handler::Sns',
  'arguments' => {
    access_key: '***AMAZON-KEY***',
    secret_key: '***AMAZON-SECRET***',
    topic_arn: 'arn:aws:sns:***'
  }.map { |k, v| "#{k}: #{v.inspect}" }
}]
```

### Usage with Amazon IAM Roles

If you are using AWS [IAM roles](http://docs.aws.amazon.com/IAM/latest/UserGuide/WorkingWithRoles.html) with your server, probably you only need to specify the `topic_arn` parameter. A few simple examples:

#### IAM Roles Method 1: in the Chef Config File

You can install the RubyGem and configure Chef to use it:

    $ gem install chef-handler-sns

Then add to the configuration (`/etc/chef/solo.rb` for chef-solo or `/etc/chef/client.rb` for chef-client):

```ruby
require 'chef/handler/sns'

exception_handlers << Chef::Handler::Sns.new(
  topic_arn: 'arn:aws:sns:us-east-1:12341234:MyTopicName'
)
```

#### IAM Roles Method 2: in a Recipe with the `chef_handler` LWRP

Use the [`chef_handler` LWRP](https://supermarket.chef.io/cookbooks/chef_handler), creating a recipe with the following:

```ruby
# Install the `chef-handler-sns` RubyGem during the compile phase
chef_gem 'chef-handler-sns' do
  compile_time true # Only for Chef 12
end

# Then activate the handler with the `chef_handler` LWRP
chef_handler 'Chef::Handler::Sns' do
  source 'chef/handler/sns'
  arguments topic_arn: 'arn:aws:sns:us-east-1:12341234:MyTopicName'
  supports exception: true
  action :enable
end
```

### IAM Roles Method 3: Using the `chef-client` Cookbook

You can also use the `node['chef_client']['config']` attribute of the [`chef-client`](https://github.com/chef-cookbooks/chef-client/tree/v4.3.2#start-report-exception-handlers) cookbook:

```ruby
node.default['chef_client']['config']['exception_handlers'] = [{
  'class' => 'Chef::Handler::Sns',
  'arguments' => ['topic_arn: "arn:aws:sns:***"']
}]
```

#### OpsWorks: Filter Notifications by Activity

An optional array of OpsWorks activities can be supplied. If the array is set, notifications will
only be triggered for the activities in the array, everything else will be discarded.

```ruby
argument_array = [
  filter_opsworks_activities: %w(deploy configure)
]
```

## Handler Configuration Options

The following options are available to configure the handler:

* `access_key` - AWS access key (required, but will try to read it from Ohai with IAM roles).
* `secret_key` - AWS secret key (required, but will try to read it from Ohai with IAM roles).
* `token` - AWS security token (optional, read from Ohai with IAM roles). Set to `false` to disable the token detected by Ohai.
* `topic_arn` - AWS topic ARN name (required).
* `message_structure` - Set this option to `json` if you want to send a different message for each protocol. You must set your [message body template](#body_template-configuration-option) properly. Valid value: `json` (optional).
* `region` - AWS region (optional).
* `subject` - Message subject string in erubis format (optional).
* `body_template` - Full path of an erubis template file to use for the message body (optional).
* `filter_opsworks_activities` - An array of OpsWorks activities to be triggered with (optional). When set, everything else will be discarded.

**Note:** When the machine has an IAM role, will try to read the credentials from Ohai. So in the best case, you only need to specify the `topic_arn`.

### `subject` Configuration Option

Here is an example of the `subject` configuration option using the ruby configuration file (`solo.rb` or `client.rb`):

```ruby
sns_handler.subject(
  "Chef-run: <%= node.name %> - <%= run_status.success? ? 'ok' : 'error' %>"
)
```

Using the [`chef_handler` LWRP](https://supermarket.chef.io/cookbooks/chef_handler):

```ruby
argument_array = [
  access_key: '***AMAZON-KEY***',
  secret_key: '***AMAZON-SECRET***',
  topic_arn: 'arn:aws:sns:***',
  subject:
    "Chef-run: <%= node.name %> - <%= run_status.success? ? 'ok' : 'error' %>"
  # [...]
]

chef_handler 'Chef::Handler::Sns' do
  # [...]
  arguments argument_array
end
```

The following variables are accessible inside the template:

* `start_time` - The time the chef run started.
* `end_time` - The time the chef run ended.
* `elapsed_time` - The time elapsed between the start and finish of the chef run.
* `run_context` - The Chef::RunContext object used by the chef run.
* `exception` - The uncaught Exception that terminated the chef run, or nil if the run completed successfully.
* `backtrace` - The backtrace captured by the uncaught exception that terminated the chef run, or nil if the run completed successfully.
* `node` - The Chef::Node for this client run.
* `all_resources` - An Array containing all resources in the chef-run resource collection.
* `updated_resources` - An Array containing all resources that were updated during the chef run.
* `success?` - Was the chef run successful? True if the chef run did not raise an uncaught exception.
* `failed?` - Did the chef run fail? True if the chef run raised an uncaught exception.

### `body_template` Configuration Option

This configuration option needs to contain the full path of an Erubis template. For example:

```ruby
# recipe 'myapp::sns_handler'

cookbook_file 'chef_handler_sns_body.erb' do
  path '/tmp/chef_handler_sns_body.erb'
  # [...]
end

argument_array = [
  access_key: '***AMAZON-KEY***',
  secret_key: '***AMAZON-SECRET***',
  topic_arn: 'arn:aws:sns:***',
  body_template: '/tmp/chef_handler_sns_body.erb'
  # [...]
]

chef_handler 'Chef::Handler::Sns' do
  # [...]
  arguments argument_array
end
```

```erb
<%# file 'myapp/files/default/chef_handler_sns_body.erb' %>

Node Name: <%= node.name %>
<% if node.attribute?('fqdn') -%>
Hostname: <%= node.fqdn %>
<% end -%>

Chef Run List: <%= node.run_list.to_s %>
Chef Environment: <%= node.chef_environment %>

<% if node.attribute?('ec2') -%>
Instance Id: <%= node.ec2.instance_id %>
Instance Public Hostname: <%= node.ec2.public_hostname %>
Instance Hostname: <%= node.ec2.hostname %>
Instance Public IPv4: <%= node.ec2.public_ipv4 %>
Instance Local IPv4: <%= node.ec2.local_ipv4 %>
<% end -%>

Chef Client Elapsed Time: <%= elapsed_time.to_s %>
Chef Client Start Time: <%= start_time.to_s %>
Chef Client Start Time: <%= end_time.to_s %>

<% if exception -%>
Exception: <%= run_status.formatted_exception %>
Stacktrace:
<%= Array(backtrace).join("\n") %>

<% end -%>
```

See the [subject](#subject) documentation for more details on the variables accessible inside the template.

If you set `message_structure` to `json`, the body template must:

* be a syntactically valid JSON; and
* contain at least a top-level JSON key of `default` with a value that is a string.

You can define other top-level keys that define the message you want to send to a specific transport protocol (e.g., "http").
```erb
<%# file 'myapp/files/default/chef_handler_sns_body.erb' %>
{
"default": "Message body text here.",
"email": "Message body text here.",
"http": "Message body text here."
}
```
See the [AWS SNS](http://docs.aws.amazon.com/sns/latest/api/API_Publish.html#API_Publish_RequestParameters) documentation for more details on SNS message format.

## IAM Role Credentials from Ohai

IAM Role information and credentials are gathered from Ohai by default if they exists.

No aditional Ohai plugin is required. This is natively supported by Ohai since version `6.16.0` ([OHAI-400](https://tickets.opscode.com/browse/OHAI-400)).

These are the used Ohai attributes:

```
ec2
├── placement_availability_zone: region is set from here.
└── iam
    └── security-credentials
        └── IAMRoleName
            ├── AccessKeyId
            ├── SecretAccessKey
            └── Token
```

## Testing

See [TESTING.md](https://github.com/zuazo/chef-handler-sns/blob/master/TESTING.md).

## Contributing

Please do not hesitate to [open an issue](https://github.com/zuazo/chef-handler-sns/issues/new) with any questions or problems.

See [CONTRIBUTING.md](https://github.com/zuazo/chef-handler-sns/blob/master/CONTRIBUTING.md).

## TODO

See [TODO.md](https://github.com/zuazo/chef-handler-sns/blob/master/TODO.md).

## License and Author

|                      |                                          |
|:---------------------|:-----------------------------------------|
| **Author:**          | [Xabier de Zuazo](https://github.com/zuazo) (<xabier@zuazo.org>)
| **Contributor:**     | [Florian Holzhauer](https://github.com/fh)
| **Contributor:**     | [Michael Hobbs](https://github.com/michaelshobbs)
| **Contributor:**     | [Hugo Lopes Tavares](https://github.com/hltbra)
| **Contributor:**     | [Dmitry Averkiev](https://github.com/d-averkiev)
| **Copyright:**       | Copyright (c) 2015 Xabier de Zuazo
| **Copyright:**       | Copyright (c) 2013-2014 Onddo Labs, SL.
| **License:**         | Apache License, Version 2.0

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
    


CHANGELOG.md
# CHANGELOG for chef-handler-sns

This file is used to list changes made in each version of `chef-handler-sns`.

## 2.1.0 (2017-02-18)

* Fix end time message typo in *body.erb* ([issue #8](https://github.com/zuazo/chef-handler-sns/pull/8), thanks [Hugo Lopes Tavares](https://github.com/hltbra)).
* Add `:message_structure` option support (issues [#9](https://github.com/zuazo/chef-handler-sns/issues/9) and [#10](https://github.com/zuazo/chef-handler-sns/issues/10), thanks [Dmitry Averkiev](https://github.com/d-averkiev)).
* Fix some Chef warnings.
* Fix some Ruby warnings.
* Update RuboCop to version `0.47`.

## 2.0.0 (2015-12-28)

### Breaking Changes on v2.0.0

* Drop Ruby `< 2` support.
* Update AWS SDK to version `2`.

### Fixes on v2.0.0

* Improve wrong encoding support (issues [#6](https://github.com/zuazo/chef-handler-sns/pull/6) and [#7](https://github.com/zuazo/chef-handler-sns/pull/7), thanks [Michael Hobbs](https://github.com/michaelshobbs)).
* Limit subject and body size properly.

### Improvements on v2.0.0

* Fix all RuboCop offenses.
* Document all the code.

### Documentation Changes on v2.0.0

* Update contact information and links after migration.
* Update chef links to use *chef.io* domain.
* README:
 * Multiple fixes, improvements and added some examples.
 * Split README file into multiple files.
 * Add GitHub and License badges.

### Changes on Tests on v2.0.0

* Tests clean up.
* Fix assertion arguments order.
* Travis CI: Test against Ruby `2.2` and Chef `12`.
* Integrate tests with `should_not` gem.

## 1.2.0 (2014-07-04)

* README:
 * Added Method 2.2: LWRP inside OpsWorks (related to [issue #4](https://github.com/zuazo/chef-handler-sns/issues/4)).
 * Add a note in the method 2 about convergence phase (related to [issue #3](https://github.com/zuazo/chef-handler-sns/issues/3)).
 * Some capitals fixed.
 * Added ohai credentials explanation (related to [cookbook issue #1](https://github.com/zuazo/chef_handler_sns-cookbook/issues/1)).
 * *filter_opsworks_activities* documentation improved.
 * Use [shields.io](http://shields.io/) badges.
 * *Contributing* section completed.
* Fixed *AWS::SNS::Errors::InvalidClientTokenId* when using version 1.1.0 ([issue #5](https://github.com/zuazo/chef-handler-sns/issues/5), thanks [Michael Hobbs](https://github.com/michaelshobbs) for reporting and testing).

## 1.1.0 (2014-03-17)

* Filter Notifications by Opsworks Activity ([issue #2](https://github.com/zuazo/chef-handler-sns/pull/2), thanks [Florian Holzhauer](https://github.com/fh))

## 1.0.0 (2014-02-19)

* Replaced `right_aws` dependency with `aws-sdk` gem.
  * Update the tests and the README.
* Add security tokens support.
* IAM roles support ([issue #1](https://github.com/zuazo/chef-handler-sns/issues/1), thanks [Michael T. Halligan](https://github.com/mhalligan) for reporting).
  * Read AWS credentials from Ohai (IAM role) by default.
* Gemfile: add_development_dependency for Ruby 1.8 compatibility now checks ruby version.
* Travis: added Ruby 2.1 to the tests.
* LICENSE: brackets replaced by copyright owner.
* Added Coverall badge again.
* Multiple README improvements and fixes.

## 0.2.6 (2013-11-15)

* Gemspec: added license.

## 0.2.5 (2013-11-01)

A Bug fix release:
* Multiple fixes in the README examples.
* Removed coveralls link from the README, sends erroneus reports.
* Added `.coveralls.yml` file.
* Removed the needless `require 'rubygems'` from the tests.
* Added `simplecov` gem for coverage tests.
* Force `mime-type` version to `~> 1.0` to fix the tests on some cases (travis related).
* Homepage changed to *GitHub Pages*.



test/helper.rb
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Xabier de Zuazo
# Copyright:: Copyright (c) 2014 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))

require 'simplecov'
if ENV['TRAVIS'] && RUBY_VERSION >= '2.1.0'
  require 'coveralls'
  SimpleCov.formatter = Coveralls::SimpleCov::Formatter
end
SimpleCov.start

gem 'minitest' # Ensures you're using the gem, and not the built in MT
require 'minitest/autorun'
require 'minitest/stub_const'
require 'mocha/setup'
require 'chef/handler/sns'
require 'should_not/minitest'


test/test_chef_handler_sns_config.rb
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Xabier de Zuazo
# Copyright:: Copyright (c) 2014 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'helper'
require 'chef/exceptions'

class SnsConfig
  include Chef::Handler::Sns::Config
end

class ChefFakeOhai < Chef::Handler::Sns::Config::Ohai
  def intialize; end
end

# rubocop:disable Metrics/BlockLength
describe Chef::Handler::Sns::Config do
  let(:node) do
    node = Chef::Node.new
    node.name('test')
    node
  end
  let(:sns_config) { SnsConfig.new }
  let(:config_params) do
    {
      access_key: '***AMAZON-KEY***',
      secret_key: '***AMAZON-SECRET***',
      token: '***AMAZON-TOKEN***',
      topic_arn: 'arn:aws:sns:***'
    }
  end

  it 'reads the configuration options on config initialization' do
    sns_config.config_init(config_params)

    assert_equal config_params[:access_key], sns_config.access_key
    assert_equal config_params[:secret_key], sns_config.secret_key
  end

  it 'is able to change configuration options using method calls' do
    sns_config.access_key(config_params[:access_key])
    sns_config.secret_key(config_params[:secret_key])

    assert_equal config_params[:access_key], sns_config.access_key
    assert_equal config_params[:secret_key], sns_config.secret_key
  end

  [:access_key, :secret_key, :topic_arn].each do |required|
    it "throws an exception when '#{required}' required field is not set" do
      config_params.delete(required)
      config_params.each { |key, value| sns_config.send(key, value) }

      assert_raises(Chef::Exceptions::ValidationFailed) do
        sns_config.config_check
      end
    end
  end

  [
    :access_key, :secret_key, :region, :token, :topic_arn, :message_structure
  ].each do |option|
    it "accepts string values in '#{option}' option" do
      sns_config.send(option, 'test')
    end

    it "sets '#{option}' option correctly" do
      sns_config.send(option, 'test')

      assert_equal 'test', sns_config.send(option)
    end

    [true, 25, Object.new].each do |bad_value|
      it "throws and exception wen '#{option}' option is set to #{bad_value}" do
        assert_raises(Chef::Exceptions::ValidationFailed) do
          sns_config.send(option, bad_value)
        end
      end
    end
  end

  it 'accepts false value to reset the token' do
    sns_config.token(false)
    assert_equal false, sns_config.token
  end

  it 'throws an exception when the body template file does not exist' do
    sns_config.body_template('/tmp/nonexistent-template.erb')
    ::File.stubs(:exist?).with(sns_config.body_template).returns(false)

    assert_raises(Chef::Exceptions::ValidationFailed) do
      sns_config.config_check
    end
  end

  describe 'config_init' do
    it 'accepts valid config options' do
      option = :access_key
      Chef::Log.expects(:warn).never

      sns_config.config_init(option => 'valid')
    end

    it 'does not accept invalid config options' do
      option = :invalid_option
      assert !sns_config.respond_to?(option)
      Chef::Log.expects(:warn).once

      sns_config.config_init(option => 'none')
    end

    it 'does not accept config options starting by "config_"' do
      option = :config_check
      assert sns_config.respond_to?(option)
      Chef::Log.expects(:warn).once

      sns_config.config_init(option => 'exists but not configurable')
    end
  end

  describe 'config_check' do
    it 'calls #config_from_ohai method' do
      sns_config.access_key(config_params[:access_key])
      sns_config.secret_key(config_params[:secret_key])
      sns_config.topic_arn(config_params[:topic_arn])
      sns_config.expects(:config_from_ohai).once

      sns_config.config_check(node)
    end
  end

  describe 'config_from_ohai' do
    before do
      @fake_ohai = ChefFakeOhai.new(node)
      Chef::Handler::Sns::Config::Ohai.stubs(:new).returns(@fake_ohai)
    end

    it 'creates Config::Ohai object' do
      Chef::Handler::Sns::Config::Ohai.expects(:new).once.returns(@fake_ohai)

      sns_config.config_from_ohai(node)
    end

    [
      :access_key,
      :secret_key,
      :token
    ].each do |method|
      it "calls Config::Ohai##{method} method" do
        @fake_ohai.expects(method).once

        sns_config.config_from_ohai(node)
      end
    end
  end
end
# rubocop:enable Metrics/BlockLength


lib/chef/handler/sns/version.rb
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Xabier de Zuazo
# Copyright:: Copyright (c) 2014 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class Chef
  class Handler
    #
    # Chef Handler SNS main class.
    #
    # A simple Chef report handler that reports status of a Chef run through
    # [Amazon SNS](http://aws.amazon.com/sns/),
    # [including IAM roles support](#usage-with-amazon-iam-roles).
    #
    class Sns < ::Chef::Handler
      #
      # chef-handler-sns Ruby Gem version.
      #
      VERSION = '2.2.0.dev'.freeze
    end
  end
end


lib/chef/handler/sns.rb
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Xabier de Zuazo
# Copyright:: Copyright (c) 2014 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/handler'
require 'chef/handler/sns/config'
require 'aws-sdk'
require 'erubis'

class Chef
  #
  # Chef report handlers.
  #
  class Handler
    #
    # Chef Handler SNS main class.
    #
    # A simple Chef report handler that reports status of a Chef run through
    # [Amazon SNS](http://aws.amazon.com/sns/),
    # [including IAM roles support](#usage-with-amazon-iam-roles).
    #
    class Sns < ::Chef::Handler
      #
      # Include {Config.config_init} and {Config.config_check} methods.
      #
      include ::Chef::Handler::Sns::Config

      #
      # Constructs a new `Sns` object.
      #
      # @example `/etc/chef/client.rb` Configuration Example
      #   require 'chef/handler/sns'
      #   sns_handler = Chef::Handler::Sns.new
      #   sns_handler.access_key '***AMAZON-KEY***'
      #   sns_handler.secret_key '***AMAZON-SECRET***'
      #   sns_handler.topic_arn 'arn:aws:sns:***'
      #   sns_handler.region 'us-east-1' # optional
      #   exception_handlers << sns_handler
      #
      # @example `/etc/chef/client.rb` Example Using a Hash for Configuration
      #   require 'chef/handler/sns'
      #   exception_handlers << Chef::Handler::Sns.new(
      #     access_key: '***AMAZON-KEY***',
      #     secret_key: '***AMAZON-SECRET***',
      #     topic_arn: 'arn:aws:sns:***',
      #     region: 'us-east-1' # optional
      #   )
      #
      # @example `/etc/chef/client.rb` Using IAM Roles
      #   require 'chef/handler/sns'
      #   exception_handlers << Chef::Handler::Sns.new(
      #     topic_arn: 'arn:aws:sns:us-east-1:12341234:MyTopicName'
      #   )
      #
      #
      # @example Using the `chef_handler` Cookbook
      #   # Install the `chef-handler-sns` RubyGem during the compile phase
      #   chef_gem 'chef-handler-sns' do
      #     compile_time true # Only for Chef 12
      #   end
      #   # Then activate the handler with the `chef_handler` LWRP
      #   chef_handler 'Chef::Handler::Sns' do
      #     source 'chef/handler/sns'
      #     arguments(
      #       access_key: '***AMAZON-KEY***',
      #       secret_key: '***AMAZON-SECRET***',
      #       topic_arn: 'arn:aws:sns:***'
      #     )
      #     supports exception: true
      #     action :enable
      #   end
      #
      # @example Using the `chef-client` Cookbook
      #   node.default['chef_client']['config']['exception_handlers'] = [{
      #     'class' => 'Chef::Handler::Sns',
      #     'arguments' => {
      #       access_key: '***AMAZON-KEY***',
      #       secret_key: '***AMAZON-SECRET***',
      #       topic_arn: 'arn:aws:sns:***'
      #     }.map { |k, v| "#{k}: #{v.inspect}" }
      #   }]
      #
      # @param config [Hash] Configuration options.
      #
      # @option config [String] :access_key AWS access key (required, but will
      #   try to read it from Ohai with IAM roles).
      # @option config [String] :secret_key AWS secret key (required, but will
      #   try to read it from Ohai with IAM roles).
      # @option config [String] :token AWS security token (optional, read from
      #   Ohai with IAM roles). Set to `false` to disable the token detected by
      #   Ohai.
      # @option config [String] :topic_arn AWS topic ARN name (required).
      # @option config [String] :region AWS region (optional).
      # @option config [String] :subject Message subject string in erubis
      #   format (optional).
      # @option config [String] :body_template Full path of an erubis template
      #   file to use for the message body (optional).
      # @option config [Array] :filter_opsworks_activities An array of
      #   OpsWorks activities to be triggered with (optional). When set,
      #   everything else will be discarded.
      #
      # @api public
      #
      def initialize(config = {})
        Chef::Log.debug("#{self.class} initialized.")
        config_init(config)
      end

      #
      # Send a SNS report message.
      #
      # This is called by Chef internally.
      #
      # @return void
      #
      # @api public
      #
      def report
        config_check(node)
        return unless allow_publish(node)
        sns.publish(
          topic_arn: topic_arn,
          message: sns_body,
          subject: sns_subject,
          message_structure: message_structure
        )
      end

      protected

      #
      # Checks if the message will be published based in configured OpsWorks
      # activities.
      #
      # @param node [Chef::Node] Chef Node that contains the activities.
      #
      # @return [Boolean] Whether the message needs to be sent.
      #
      # @api private
      #
      def allow_publish(node)
        return true if filter_opsworks_activity.nil?

        if node.attribute?('opsworks') &&
           node['opsworks'].attribute?('activity')
          return filter_opsworks_activity.include?(node['opsworks']['activity'])
        end

        Chef::Log.debug(
          'You supplied opsworks activity filters, but node attr was not '\
          'found. Returning false'
        )
        false
      end

      #
      # Returns the {Aws::SNS} object used to send the messages.
      #
      # @return [Aws::SNS::Client] The SNS client.
      def sns
        @sns ||= begin
          params = {
            access_key_id: access_key,
            secret_access_key: secret_key,
            logger: Chef::Log
          }
          params[:region] = region if region
          params[:session_token] = token if token
          Aws::SNS::Client.new(params)
        end
      end

      #
      # Fixes or forces the correct encoding of strings.
      #
      # Replaces wrong characters with `'?'`s.
      #
      # @param o [String, Object] The string to fix.
      # @param encoding [String] The encoding to use.
      #
      # @return [String] The message fixed.
      #
      # @api private
      #
      def fix_encoding(o, encoding)
        encode_opts = { invalid: :replace, undef: :replace, replace: '?' }

        return o.to_s.encode(encoding, encode_opts) if RUBY_VERSION >= '2.1.0'
        # Fix ArgumentError: invalid byte sequence in UTF-8 (issue #7)
        o.to_s.encode(encoding, 'binary', encode_opts)
      end

      #
      # Fixes the encoding of SNS subjects.
      #
      # @param o [String, Object] The subject to fix.
      #
      # @return [String] The message fixed.
      #
      # @api private
      #
      def fix_subject_encoding(o)
        fix_encoding(o, 'ASCII')
      end

      #
      # Fixes the encoding of SNS bodies.
      #
      # @param o [String, Object] The body to fix.
      #
      # @return [String] The message fixed.
      #
      # @api private
      #
      def fix_body_encoding(o)
        fix_encoding(o, 'UTF-8')
      end

      #
      # Returns the SNS subject used by default.
      #
      # @return [String] The SNS subject.
      #
      # @api private
      #
      def default_sns_subject
        chef_client = Chef::Config[:solo] ? 'Chef Solo' : 'Chef Client'
        status = run_status.success? ? 'success' : 'failure'
        fix_subject_encoding("#{chef_client} #{status} in #{node.name}"[0..99])
      end

      #
      # Limits the size of a UTF-8 string in bytes without breaking it.
      #
      # Based on http://stackoverflow.com/questions/12536080/
      # ruby-limiting-a-utf-8-string-by-byte-length
      #
      # @param str [String] The string to limit.
      # @param size [Integer] The string size in bytes.
      #
      # @return [String] The final string.
      #
      # @note This code does not work properly on Ruby `< 2.1`.
      #
      # @api private
      #
      def limit_utf8_size(str, size)
        # Start with a string of the correct byte size, but with a possibly
        # incomplete char at the end.
        new_str = str.byteslice(0, size)

        # We need to force_encoding from utf-8 to utf-8 so ruby will
        # re-validate (idea from halfelf).
        until new_str[-1].force_encoding('utf-8').valid_encoding?
          # Remove the invalid char
          new_str = new_str.slice(0..-2)
        end
        new_str
      end

      #
      # Generates the SNS subject.
      #
      # @return [String] The subject string.
      #
      # @api private
      #
      def sns_subject
        return default_sns_subject unless subject
        context = self
        eruby = Erubis::Eruby.new(fix_subject_encoding(subject))
        fix_subject_encoding(eruby.evaluate(context))[0..99]
      end

      #
      # Generates the SNS body.
      #
      # @return [String] The body string.
      #
      # @api private
      #
      def sns_body
        template = IO.read(body_template ||
          "#{File.dirname(__FILE__)}/sns/templates/body.erb")
        context = self
        eruby = Erubis::Eruby.new(fix_body_encoding(template))
        body = fix_body_encoding(eruby.evaluate(context))
        limit_utf8_size(body, 262_144)
      end
    end
  end
end
